clients/src/main/java/org/apache/kafka/clients/consumer/AbstractPartitionAssignor.java

```diff
@@ -0,0 +1,331 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.types.ArrayOf;
+import org.apache.kafka.common.protocol.types.Field;
+import org.apache.kafka.common.protocol.types.GenericType;
+import org.apache.kafka.common.protocol.types.Schema;
+import org.apache.kafka.common.protocol.types.SchemaException;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.protocol.types.Type;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Abstract assignor implementation used by default assignors shipped with Kafka. The embedded
+ * format of assignments and subscriptions are as follows:
+ *
+ * <pre>
+ * Subscription => Version Topics
+ *   Version       => int16
+ *   Topics        => [String]
+ *
+ * Assignment => Version TopicPartitions
+ *   Version         => int16
+ *   TopicPartitions => [Topic Partitions]
+ *     Topic         => String
+ *     Partitions    => [int32]
+ * </pre>
+ *
+ * This assignor assumes that any changes to these formats will be compatible with this format.
+ * In other words, newer versions can be parsed using the old format. This generally means that
+ * fields can be added to the end of the structures, but no existing fields can be removed or reordered.
+ */
+public abstract class AbstractPartitionAssignor implements PartitionAssignor<AbstractPartitionAssignor.Subscription, AbstractPartitionAssignor.Assignment> {
+    private static final Logger log = LoggerFactory.getLogger(AbstractPartitionAssignor.class);
+
+    public static final String VERSION_KEY_NAME = "version";
```

Comment:
@hachikuji  I was originally mainly think about code simplicity, not about tooling connivence much. I am wondering for CopyCat, is a topics field always necessary and representative of the resource partitions?

I get @ewencp 's point that enforcing the whole protocol to be restricted to Kafka structures would be less flexible in terms of compatibility, and if we only do that by extracting ```topics``` from subscriptions while keep ```userData``` and  ```assignment``` still generic types then it does not buy us much regarding code simplicity either. So I am now OK with the current approach of keeping ```GenericType```. The only other comment is as below: we probably can save the AbstractPartitionAssignor as its only ```assign``` function is more like a helper function, hence putting the schemas in a ```ConsumerProtocol``` class along with the helper functions, and only keep one ```PartitionAssignor``` interface for users to instantiate.